### PR TITLE
Replace keychain with go keychain

### PR DESCRIPTION
**Pull request by lox**
I've been meaning to do this for ages, but finally got around to putting in the PR's to https://github.com/keybase/go-keychain that will let us entirely replace our long-ago-forked custom keychain interaction code. So much joy in this diff.

https://github.com/keybase/go-keychain/pull/12
https://github.com/keybase/go-keychain/pull/13

### CODE REVIEW

File: keyring/keychain.go [hunk 1]
```diff
@@ -2,27 +2,15 @@
...
-	"path/filepath"
-	"unicode/utf8"
-	"unsafe"
 
-	"github.com/mitchellh/go-homedir"
+	gokeychain "github.com/keybase/go-keychain"
+)
+
+const (
+	keychainAccessGroup = "ACE1234DEF.com.99designs.aws-vault"
```
Comment (pda) [inline keyring/keychain.go]: I don't fully understand what this is, but looks like it maps to [`kSecAttrAccessGroup`](https://developer.apple.com/documentation/security/ksecattraccessgroup) which doesn't sound necessary in this use-case, and `ACE1234DEF` looks suspiciously like sample data.

> Access groups are used to share keychain items among two or more apps.

It's being passed to `gokeychain.NewGenericPassword()` which calls [`SetAccessGroup()`](https://github.com/keybase/go-keychain/blob/69c518e5bfff89ad62db81a185634d3aeac3c453/keychain.go#L226) which calls [`k.SetString(AccessGroupKey, ag)`](https://github.com/keybase/go-keychain/blob/69c518e5bfff89ad62db81a185634d3aeac3c453/keychain.go#L188) which treats empty string as unset. So perhaps we should just pass an empty string to `gokeychain.NewGenericPassword()`? Or is this actually doing something?

Comment (lox) [inline keyring/keychain.go]: Yup, agreed. It's for sharing items between a family of apps, which we don't need.

File: keyring/keychain.go [hunk 2]
```diff
@@ -37,483 +20,131 @@ func init() {
 			name = "login"
 		}
 
-		home, err := homedir.Dir()
-		if err != nil {
-			return nil, err
-		}
-
 		return &keychain{
-			path:    filepath.Join(home, "/Library/Keychains/"+name+".keychain"),
```
Comment (lox) [inline keyring/keychain.go]: This is one change that probably needs testing. The new code uses simply name +".keychain", vs the old one that references the full path. It works fine on my machine which is 10.12.6.

File: keyring/keychain.go [hunk 3]
```diff
@@ -37,483 +20,131 @@ func init() {
...
-	queryDict := mapToCFDictionary(query)
-	defer C.CFRelease(C.CFTypeRef(queryDict))
+	newItem := gokeychain.NewItem()
+	newItem.SetSecClass(gokeychain.SecClassGenericPassword)
+	newItem.SetService(k.service)
+	newItem.SetAccount(item.Key)
+	newItem.SetLabel(item.Label)
+	newItem.SetDescription(item.Description)
+	newItem.SetData(item.Data)
+	newItem.SetSynchronizable(gokeychain.SynchronizableNo)
```
Comment (lox) [inline keyring/keychain.go]: I like that this is more specific. I wonder what the old default was.

File: keyring/keychain.go [hunk 4]
```diff
@@ -37,483 +20,131 @@ func init() {
...
-
-	if err == errDuplicateItem {
-		if err = k.Remove(item.Key); err != nil {
-			return err
-		}
-		err = newKeychainError(C.SecItemAdd(queryDict, nil))
-	}
-
-	return err
+	return gokeychain.AddItem(item)
```
Comment (mtibben) [inline keyring/keychain.go]: I'm getting a compile error?
```
$ go build github.com/99designs/aws-vault
# github.com/99designs/aws-vault/keyring
keyring/keychain.go:85: cannot use item (type Item) as type keychain.Item in argument to keychain.AddItem
```

Comment (mtibben) [inline keyring/keychain.go]: `s/item/newItem/` works

**Comment by mtibben** [comment]
All looks good to me

**Comment by lox** [comment]
We really need some sort of CI for this project.

**Comment by mtibben** [comment]
Actually, I have spotted a difference in behaviour. When creating a keychain, v3.7.1 prompts for a password. This version does not. Is that intentional?

**Comment by lox** [comment]
Nope, not intentional.

**Comment by lox** [comment]
Good spotting.

File: keyring/keychain.go [hunk 5]
```diff
@@ -37,483 +20,131 @@ func init() {
...
-	}
-	return C.CFArrayCreate(nil, valuesPointer, C.CFIndex(numValues), &C.kCFTypeArrayCallBacks)
-}
 
-func _CFArrayToArray(cfArray C.CFArrayRef) (a []C.CFTypeRef) {
-	count := C.CFArrayGetCount(cfArray)
-	if count > 0 {
-		a = make([]C.CFTypeRef, count)
-		C.CFArrayGetValues(cfArray, C.CFRange{0, count}, (*unsafe.Pointer)(&a[0]))
+	if k.passphrase != "" {
```
Comment (lox) [inline keyring/keychain.go]: 🤦‍♂️

**Comment by lox** [comment]
Verified this works.

**Comment by joho** [comment]
Hey, I was just going through documenting/tidying stuff in the keyring extraction I did.

Looks like https://github.com/keybase/go-keychain/pull/13 never merged. I'm wondering if there's a bug hidden somewhere and it's no longer doing what we think it does.

**Comment by lox** [comment]
Nah, those folks just take a while to respond sometimes.

**Pull request merged**